1_fetch/src/fetch_precip_values.R

```diff
@@ -0,0 +1,77 @@
+
+fetch_precip_values <- function(ind_file, precip_spatial_ind, dates) {
+
+  nc <- ncdf4::nc_open("https://cida.usgs.gov/thredds/dodsC/stageiv_combined")
+
+  precip_spatial <- readRDS(sc_retrieve(precip_spatial_ind))
+
+  t_vals <- get_time_nc(nc$dim$time)
+
+  start_date <- as.Date(dates$start)
+  end_date <- as.Date(dates$end)
+
+  x_inds <- seq(min(precip_spatial$x), max(precip_spatial$x), 1)
+  y_inds <- seq(min(precip_spatial$y), max(precip_spatial$y), 1)
+  t_inds <- which(t_vals > start_date & t_vals < end_date)
+
+  # ensures we make our request in the right axis order.
+  dimid_order <- match(nc$var$Total_precipitation_surface_1_Hour_Accumulation$dimids,
+                       c(nc$dim$x$id, nc$dim$y$id, nc$dim$time$id))
+
+  precip <- array(dim = c(length(x_inds), length(y_inds), length((t_inds))))
+  t_counter <- 1
+
+  for(step in seq(min(t_inds), max(t_inds), 1)) {
+    precip[,,t_counter] <-
+      ncdf4::ncvar_get(nc, nc$var$Total_precipitation_surface_1_Hour_Accumulation,
+                       start <- c(min(x_inds),
+                                  min(y_inds),
+                                  step)[dimid_order],
+                       count <- c(length(x_inds),
+                                  length(y_inds),
+                                  1)[dimid_order])
+    t_counter <- t_counter + 1
+    message(paste(t_counter, length(t_inds)))
+  }
+
+  precip <- arrayhelpers::array2df(precip)
+
+  x_inds <- data.frame(d1 = seq_len(length(x_inds)), x = x_inds)
+  y_inds <- data.frame(d2 = seq_len(length(y_inds)), y = y_inds)
+  t_inds <- data.frame(d3 = seq_len(length(t_inds)), t = t_inds)
+
+  precip <- precip %>%
+    left_join(x_inds, by = "d1") %>%
+    left_join(y_inds, by = "d2") %>%
+    left_join(t_inds, by = "d3") %>%
+    inner_join(sf::st_set_geometry(precip_spatial, NULL),
+               by = c("x", "y")) %>%
+    left_join(data.frame(time = t_vals,
+                         t = seq_len(length(t_vals))),
+              by = "t") %>%
+    select(precip, time, id)
```

Comment:
Seems fine, but any reason you used a skinny data.frame instead of a 3d matrix of values? Having not played around w/ either, not sure if this one offers advantages that I am not seeing. Given all of the repeated values in the `$time` field (191 unique values in 37880266 rows), I think the data.frame format is quite a bit bigger.

Reply:
As we talked about on the phone -- this is a bit safer. And since we are joining it up to a set of spatial points, this seemed like a happy medium. I'm going to stick with this format and move on to process where I'll dump out a list of rasters with time stamp as the name of each raster.